### ISSUE TITLE
[bug] : `requirements.txt` files across projects use unpinned dependencies — notebooks silently break as packages release new major versions

### ISSUE THREAD

**Issue by divyanshim27** [abhisheks008/DL-Simplified#1148]
**Problem**

Across the 500+ project folders, `requirements.txt` files typically list packages without version pins:
tensorflow
opencv-python
scikit-learn
This means a notebook that worked correctly in 2022 with `tensorflow==2.9.0` will silently fail or produce different results in 2026 when `tensorflow==2.17.0` is installed — due to breaking API changes (e.g., Keras 3 migration in TF 2.16+).

**Impact**

Students following the repo for educational purposes run notebooks that were never tested against the installed version and encounter confusing errors with no debugging path.

**Proposed Fix**

- Update the contribution template to **require** pinned dependencies in `requirements.txt`:
tensorflow==2.13.0
opencv-python==4.8.0.76
scikit-learn==1.3.0
- Add a validation step to the proposed CI workflow (Issue above) that rejects `requirements.txt` files with unpinned top-level packages
- Add a note in `CONTRIBUTING.md`: *"Always pin your dependency versions. Run `pip freeze > requirements.txt` after testing your notebook."*

**Labels:** `bug`, `documentation`, `reproducibility`, `good first issue`

**Comment by divyanshim27** [comment]
Hey @abhisheks008   ,
Please assign it to me under GSSOC 2026.
/assign /claim

**Comment by nischay25** [comment]
Hi @abhisheks008 

I am a GSSoC'26 contributor and would like to work on this issue.

I understand the importance of dependency pinning for reproducibility, especially in educational machine learning projects. My approach would be to:

* Review the current dependency management guidelines.
* Update the contribution documentation to recommend pinned package versions.
* Add clear instructions for contributors to generate reproducible `requirements.txt` files after testing.
* Implement or update the validation workflow to detect unpinned top-level dependencies where applicable.
* Ensure the changes improve reproducibility without affecting the existing project structure.

I will keep the implementation clean, well-documented, and aligned with the repository's contribution guidelines.

Kindly assign this issue to me if it is available.

Thank you!

**Comment by karrisanthoshigayatri** [comment]
Hi @abhisheks008,

I'd like to work on this issue if it is still available. My approach is to update the contribution guidelines with dependency version pinning, add validation checks for unpinned packages in `requirements.txt`, and improve reproducibility while keeping the existing project structure unchanged.

Kindly consider assigning this issue to me if it becomes available. Thank you!
